src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpOnTypeFormattingPass.cs

```diff
@@ -146,9 +149,54 @@ public async override Task<FormattingResult> ExecuteAsync(FormattingContext cont
             var finalChanges = cleanedText.GetTextChanges(originalText);
             var finalEdits = finalChanges.Select(f => f.AsTextEdit(originalText)).ToArray();
 
+            if (context.AutomaticallyAddUsings)
+            {
+                // Because we need to parse the C# code twice for this operation, lets do a quick check to see if its even necessary
+                if (result.Edits.Any(e => e.NewText.IndexOf("using") != -1))
```

Comment:
Beautiful!

Reply:
> So the assumption is that if any usings are added all the necessary usings are added?

I'm not sure I follow exactly. The way this behaves is if any usings are edited, we'll check to see if any have been added.

> What about: var usingVariables = "fun";?

Yes, this will definitely have false positives. The only way around that is to semantically understand the edits, and if we could do that we wouldn't need any of this code anyway :)